### DISCUSSION TITLE
7주차: 10장 조건부 로직 간소화

### DISCUSSION THREAD

**Discussion by choisohyun** [frontend-book-study/refactoring-2#15] [리팩터링 2판]
인상깊었던 책의 내용과 인상깊었던 이유
현업 등 코드에 반영할 수 있는 방법과 생각
관련 아티클이나 이해를 돕는 자료를 첨부
새롭게 알게된 점이나 이해가 어려웠던 점

**Comment by choisohyun** [comment]
`10.5 특이 케이스 추가하기` 챕터가 인상적이었습니다. 작년에 고민을 많이 했던 케이스 중에 하나가 이와 비슷했습니다.(예시와 같은 함수가 아니라 컴포넌트 단위이긴 했습니다) 계정의 타입에 따라 일부만 다른 요소나 값을 노출해야 했습니다. 예시로 보여준 케이스들 모두 별개의 동작을 하는 객체를 생성하여 복잡도를 줄입니다. 형태는 조금 다르긴 하지만 어느정도 비슷한 결과로 만든 것 같아 다행입니다. 
나머지 챕터는 이런 류의 조건부 로직 간소화 리팩터링을 좋아하기 때문에 자연스럽게 봤던 것 같습니다!

**Comment by BEMELON** [comment]
`10.6 Assertion 추가하기` 

종종 코드에서 보이긴 했었는데, 이걸 프레임워크에서 정의한 에러 처리로 진행했어야 했는 지 고민이 됐던 부분이 있었습니다. 
책에서 이러한 내용에 크게 공감하고 추가로 배우게 됐던 것 같습니다 !
```
한편, 어서션을 남발하는 것 역시 위험하다. 나는 참이라고 생각하는 가정 모두에 어서션을 달지는 않는다. ‘반드시 참이어야 하는’ 것만 검사한다.
```

**Comment by hamo-o** [comment]
> 10.3 중첩 조건문을 보호 구문으로 바꾸기 - 조건문은 두 가지 형태로 쓰인다. 
> 1. 참인 경로와 거짓인 경로 모두 정상 동작으로 이어짐 → `if~else` 두 절에 똑같은 무게를 둔다.
> 2. 한쪽만 정상 → 보호 구문 `if` 로 “**이건 이 함수의 핵심이 아니다.** 이 일이 일어나면 무언가 조치를 취한 후 함수에서 빠져나온다.” 를 나타낸다.

코드를 작성할 때 최대한 `else`문을 지양하고 바로 `return`하는 식으로 작성했는데, 이제 어떠한 기준으로 `if~else`와 `if`를 나누어야 할지 감이 잡혔습니다. "이 부분은 이 함수의 핵심이 아니다." 라는 부분을 먼저 `return`하고, 그렇지 않고 둘 모두 정상 동작이며 양 갈래가 동일하게 중요할 때는 `else`문을 사용하는 것을 꺼리지 않아도 된다는 것을 배웠습니다.

**Comment by RAINBOWJYP** [comment]
최근 조건이 많은 조건부를 만들었는데 내가 봐도 if-then-else가 너무 길어서 보기 어렵다는 생각을 했었습니다.
조건식을 함수로 추출하는것을 보고 가독성을 해소하는데 좋은 수단이 되겠구나 라고 생각했습니다!
제일 재밌게 봤던 부분은 '특이 케이스 추가하기'였습니다. 두번 읽었는데요 (이해못한것도 있었음 ㅎㅎ)
사내 서비스중에 확인 어려운 케이스가 하나 있는데 이 기법을 활용하면 좋을 것 같았습니다.

**Comment by suojae** [comment]
1. 조건문을 작성함에 있어 핵심로직을 명확히 보이게 작성해야함을 느꼈습니다. 핵심로직을 위한 잔잔바리 케이스를 미리 앞부분에서 필터링해줌으로써 핵심로직과 잔잔바리 로직을 시각적으로 분리시킬 수 있겠다라는 생각이 들었습니다.

2. 평소 switch 문으로 분기처리하는 것을 좋아했는데 이부분을 클래스 생성 및 상속을 통해 좀더 역할과 책임에 맞는 코드작성을 할 수 있다는 것이 인상적이었습니다!
